storage/sqlite/src/lib.rs

```diff
@@ -283,19 +286,28 @@ impl Sqlite {
     }
 }
 
+/// Implements in memory database, useful for testing
+impl Default for Sqlite {
+    fn default() -> Self {
+        Self { path: None }
+    }
+}
+
 impl backend::Backend for Sqlite {
     type Impl = SqliteImpl;
 
     fn open(self, desc: DbDesc) -> storage_core::Result<Self::Impl> {
-        // Attempt to create the parent storage directory
-        if let Some(parent) = self.path.parent() {
-            std::fs::create_dir_all(parent).map_err(error::process_io_error)?;
-        } else {
-            return Err(storage_core::error::Recoverable::Io(
-                std::io::ErrorKind::NotFound,
-                "Cannot find the parent directory".to_string(),
-            )
-            .into());
+        // Attempt to create the parent storage directory if using a file
+        if let Some(ref path) = self.path {
+            if let Some(parent) = path.parent() {
+                std::fs::create_dir_all(parent).map_err(error::process_io_error)?;
+            } else {
+                return Err(storage_core::error::Recoverable::Io(
```

Comment:
I had an impression that recoverable errors are errors that can be fixed by retrying the operation. In that case this error doesn't look like a recoverable one. On the other hand we have the `DbInit` recoverable error. Can someone explain this to me? 🙃

Reply:
In my conception, recoverable errors are ones that can be handled gracefully, irrecoverable are ones where very little can be done about (e.g. out of memory, etc.). It's fine to re-categorize existing errors if the current classification does not make sense. It's probably a good idea to add another category to separate "retry-able errors" as you suggest.

Reply:
This makes sense, I am changing it to a Fatal error.

Reply:
The difference between recoverable and irrecoverable errors is simply whether we have a plan to recover. That's all. Do we know how to recover if we get error X? That's what decides it. I/O is usually an irrecoverable error, unless we know what to do and how to fix it, which quite frankly is rare. If a file doesn't work or you can't write on it, what do you really hope to do exactly? And I keep saying this: It could be a hardware failure. That's a game we shouldn't play.

I don't think we need a "retry-able errors"... that's what recoverable is for. The application should decide whether it needs to retry.

An example of a recoverable error is a transaction failing in lmdb due to size limitations. We can "try again" and expect a resize to have happened. We try a few times, then we give up and crash.

Another example is a transaction temporal serializability failure. Hence if tx A starts, then tx B, then both modify the same value. Only one of them can commit. The other should "retry" and hope no other conflicts happen.